base/loading.jl

```diff
@@ -145,38 +183,62 @@ function include_from_node1(path::AbstractString)
-    had = haskey(package_list, path)
-    if !had
-        package_locks[path] = (false, Condition())
+function evalfile(path::AbstractString, args::Vector{UTF8String}=UTF8String[])
+    return eval(Module(:__anon__),
+                Expr(:toplevel,
+                     :(const ARGS = $args),
+                     :(eval(x) = Main.Core.eval(__anon__,x)),
+                     :(eval(m,x) = Main.Core.eval(m,x)),
+                     :(Main.Base.include($path))))
+end
+evalfile(path::AbstractString, args::Vector) = evalfile(path, UTF8String[args...])
+
+function create_expr_cache(file::AbstractString, name)
+    cachepath = LOAD_CACHE_PATH[1]
+    if !isdir(cachepath)
+        mkpath(cachepath)
     end
-    package_list[path] = time()
-    tls = task_local_storage()
-    prev = pop!(tls, :SOURCE_PATH, nothing)
-    try
-        eval(Main, :(Base.include_from_node1($path)))
-    catch e
-        had || delete!(package_list, path)
-        rethrow(e)
-    finally
-        if prev != nothing
-            tls[:SOURCE_PATH] = prev
+    cachefile = abspath(cachepath, name*".ji")
+    code_object = """
+        while !eof(STDIN)
+            eval(Main, deserialize(STDIN))
         end
+        """
+    io, pobj = open(detach(setenv(`$(julia_cmd())
+            --output-ji $cachefile --output-incremental=yes
+            --startup-file=no --history-file=no
+            --eval $code_object`,
+        ["JULIA_HOME=$JULIA_HOME", "HOME=$(homedir())"])), "w")
+    serialize(io, quote
+        empty!(Base.LOAD_PATH)
+        append!(Base.LOAD_PATH, $LOAD_PATH)
+        empty!(Base.LOAD_CACHE_PATH)
+        append!(Base.LOAD_CACHE_PATH, $LOAD_CACHE_PATH)
+        empty!(Base.DL_LOAD_PATH)
+        append!(Base.DL_LOAD_PATH, $DL_LOAD_PATH)
+    end)
+    source = source_path(nothing)
+    if source !== nothing
+        serialize(io, quote
+            task_local_storage()[:SOURCE_PATH] = $(source)
+        end)
     end
-    reloaded, c = package_locks[path]
-    if !reloaded
-        package_locks[path] = (true, c)
-        notify(c, all=true)
+    serialize(io, :(Base.include($(abspath(file)))))
+    if source !== nothing
+        serialize(io, quote
+            delete!(task_local_storage(), :SOURCE_PATH)
+        end)
     end
-    nothing
+    close(io)
+    wait(pobj)
+    return cachefile
 end
 
-function evalfile(path::AbstractString, args::Vector{UTF8String}=UTF8String[])
-    return eval(Module(:__anon__),
-                Expr(:toplevel,
-                     :(const ARGS = $args),
-                     :(eval(x) = Core.eval(__anon__,x)),
-                     :(eval(m,x) = Core.eval(m,x)),
-                     :(include($path))))
+compile(mod::Symbol) = compile(string(mod))
+function compile(name::AbstractString)
+    path = find_in_node_path(name, 1)
+    path === nothing && throw(ArgumentError("$name not found in path"))
+    # spawn off a new incremental compile task from node 1
+    return remotecall_fetch(1, create_expr_cache, path, name)
```

Comment:
I think you misunderstand me; I mean that calling `compile` should simply write a file locally, on the same node where the call happens. I don't think we need to do anything distributed in `compile`; that can all be done in loading only.